File: apps/frontend/web/src/api/spaces.ts
import type { Space, SpaceData, Topic, ContentOverview, ContentType } from '@/types/space'

const API_BASE = 'http://localhost:8000'

async function fetchApi<T>(url: string, options?: RequestInit & { params?: Record<string, string | number> }): Promise<T> {
  let fullUrl = url
  
  if (options?.params) {
    const searchParams = new URLSearchParams()
    Object.entries(options.params).forEach(([key, value]) => {
      searchParams.append(key, String(value))
    })
    fullUrl = `${url}?${searchParams.toString()}`
  }
  
  const response = await fetch(`${API_BASE}${fullUrl}`, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      ...options?.headers,
    },
  })
  
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: 'Unknown error' }))
    throw new Error(error.detail || 'API error')
  }
  
  if (response.status === 204) {
    return undefined as T
  }
  
  return response.json()
}

export interface TopicInput {
  title: string
  contentOverview?: ContentOverview
  contentType?: ContentType
}

export interface PaginatedSpaces {
  spaces: Space[]
  total: number
  limit: number
  offset: number
}

export const api = {
  // Spaces
  listSpaces: (limit = 12, offset = 0) => fetchApi<PaginatedSpaces>('/api/spaces', { params: { limit, offset } }),
  
  getSpace: (slug: string) => fetchApi<SpaceData>(`/api/spaces/${slug}`),
  
  createSpace: (title: string, description: string = '') => 
    fetchApi<Space>('/api/spaces', {
      method: 'POST',
      body: JSON.stringify({ title, description }),
    }),
  
  createSpaceWithTopics: (title: string, topics: TopicInput[]) =>
    fetchApi<SpaceData>('/api/spaces/with-topics', {
      method: 'POST',
      body: JSON.stringify({ title, topics }),
    }),
  
  updateSpace: (slug: string, data: { title?: string; description?: string }) =>
    fetchApi<SpaceData>(`/api/spaces/${slug}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    }),
  
  deleteSpace: (slug: string) => 
    fetchApi<void>(`/api/spaces/${slug}`, { method: 'DELETE' }),
  
  // Topics
  createTopic: (slug: string, title: string, contentType?: ContentType) =>
    fetchApi<Topic>(`/api/spaces/${slug}/topics`, {
      method: 'POST',
      body: JSON.stringify({ title, content_type: contentType }),
    }),
}

File: apps/backend/agent-api/api/routers/spaces.py
from fastapi import APIRouter, HTTPException, Query
from typing import List, Optional, Union, Any
from pydantic import BaseModel

from api.schemas.space import (
    Space,
    SpaceCreate,
    SpaceUpdate,
    SpaceWithTopics,
    TopicCreate,
    Topic,
    PaginatedSpaces,
)
from api.services import space_service

router = APIRouter(prefix="/api/spaces", tags=["spaces"])


class TopicInputItem(BaseModel):
    title: str
    contentOverview: Optional[dict] = None
    contentType: Optional[str] = None


class SpaceWithTopicsCreate(BaseModel):
    title: str
    description: str = ""
    topics: List[Any]


@router.get("", response_model=PaginatedSpaces)
def list_spaces(limit: int = Query(10, ge=1, le=50), offset: int = Query(0, ge=0)):
    return space_service.list_spaces(limit=limit, offset=offset)


@router.post("", response_model=Space, status_code=201)
def create_space(data: SpaceCreate):
    try:
        return space_service.create_space(data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{slug}", response_model=SpaceWithTopics)
def get_space(slug: str):
    space = space_service.get_space(slug)
    if not space:
        raise HTTPException(status_code=404, detail="Space not found")
    return space


@router.put("/{slug}", response_model=SpaceWithTopics)
def update_space(slug: str, data: SpaceUpdate):
    space = space_service.update_space(slug, data)
    if not space:
        raise HTTPException(status_code=404, detail="Space not found")
    return space


@router.delete("/{slug}", status_code=204)
def delete_space(slug: str):
    if not space_service.delete_space(slug):
        raise HTTPException(status_code=404, detail="Space not found")


@router.post("/{slug}/topics", response_model=Topic, status_code=201)
def create_topic(slug: str, data: TopicCreate):
    topic = space_service.create_topic(slug, data)
    if not topic:
        raise HTTPException(status_code=404, detail="Space not found")
    return topic


@router.post("/with-topics", response_model=SpaceWithTopics, status_code=201)
def create_space_with_topics(data: SpaceWithTopicsCreate):
    try:
        return space_service.create_space_with_topics(
            data.title, data.description, data.topics
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
